### PR TITLE
Rng Generation

### PR DESCRIPTION
**Pull request by ForbiddenKJ**
A class for the API that rolls the "rarities" using the provided weights, these may need tweaked further down the line but from what I can see they look good.

Closing #3

### CODE REVIEW

File: src/api/models/rarity.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class Rarity(Enum):
+    Poor: list[float] = [50.077, 27.337, 16.916, 4.787, 0.815, 0.066, 0.002]
+    Common: list[float] = [27.738, 49.889, 16.781, 4.732, 0.777, 0.078, 0.005]
+    Uncommon: list[float] = [10.5, 23.031, 55.163, 9.665, 1.455, 0.18, 0.006]
+    Rare: list[float] = [0.0, 0.0, 11.403, 74.866, 13.362, 0.351, 0.018]
+    Epic: list[float] = [0.0, 0.0, 0.0, 85.014, 13.451, 1.465, 0.07]
+    Legendary: list[float] = [0.0, 0.0, 0.0, 10.0, 79.979, 9.509, 0.512]
+    Mythical: list[float] = [0.0, 0.0, 0.0, 0.0, 10.0, 84.932, 5.068]
```
Comment (GDWR) [inline src/api/models/rarity.py:11]: The Enum values should be 0, 1, 2, 3, 4, 5, 6
This makes it so we can store the Enum in the database easier.

This Rarity can implement a property to gain the weights. i.e.
```py

class Rarity(Enum):
    Poor = 0
    Common = 1
    Uncommon = 2
    Rare = 3
    Epic = 4
    Legendary = 5
    Mythical = 6
    
    @property
    def weights(self) -> list[float]:
            if self.value is Rarity.Poor:
                 Poor: list[float] = [50.077, 27.337, 16.916, 4.787, 0.815, 0.066, 0.002]
            ...

```

File: src/api/services/rng.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from random import choices
+from models import Rarity
+from typing import Union
+
+
+class RNG:
+    def __init__(self, weights: Union[Rarity, list[int]] = None):
```
Comment (GDWR) [inline src/api/services/rng.py:7]: int -> float

File: src/api/services/rng.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            Rarity.Poor,
+            Rarity.Common,
+            Rarity.Uncommon,
+            Rarity.Rare,
+            Rarity.Epic,
+            Rarity.Legendary,
+            Rarity.Mythical,
+        ]
+
+    def roll(self, k: int = 1) -> Rarity:
```
Comment (GDWR) [inline src/api/services/rng.py:28]: Docstring to explain what param `k` is

**Review by GDWR — CHANGES_REQUESTED** [review]
The Enum is going to be used in the database and the values really should be ints ranging from 0 - n

File: src/api/models/rarity.py [hunk 4]
```diff
@@ -2,10 +2,33 @@
...
-    Epic: list[float] = [0.0, 0.0, 0.0, 85.014, 13.451, 1.465, 0.07]
-    Legendary: list[float] = [0.0, 0.0, 0.0, 10.0, 79.979, 9.509, 0.512]
-    Mythical: list[float] = [0.0, 0.0, 0.0, 0.0, 10.0, 84.932, 5.068]
+    Poor: list[float] = 1
+    Common: list[float] = 2
+    Uncommon: list[float] = 3
+    Rare: list[float] = 4
+    Epic: list[float] = 5
+    Legendary: list[float] = 6
+    Mythical: list[float] = 7
```
Comment (GDWR) [inline src/api/models/rarity.py:11]: Type hinting is wrong

File: src/api/services/rng.py [hunk 5]
```diff
@@ -4,16 +4,16 @@
 
 
 class RNG:
-    def __init__(self, weights: Union[Rarity, list[int]] = None):
-        if weights is None:
-            self.weights = Rarity.Poor.value
+    def __init__(self, rarity_weights: Union[Rarity, list[int]] = None):
```
Comment (GDWR) [inline src/api/services/rng.py:7]: list[int] should be list[float]

Comment (GDWR) [reply src/api/services/rng.py:7]: keeping it as `weights` for the param name makes more sense too.

File: src/api/services/rng.py [hunk 6]
```diff
@@ -4,16 +4,16 @@
...
-                self.weights = weights.value
+            if isinstance(rarity_weights, Rarity):
+                self.weights = rarity_weights.weights
 
-            elif isinstance(weights, list):
-                self.weights = weights
+            elif isinstance(rarity_weights, list):
+                self.weights = rarity_weights
 
         self.rewards: list[Rarity] = [
```
Comment (GDWR) [inline src/api/services/rng.py:18]: Pretty sure you can just `list(Rarity)` to make the enum a list

**Review by GDWR — CHANGES_REQUESTED** [review]

File: src/api/services/rng.py (hunk 6 above)
Comment (ForbiddenKJ) [reply src/api/services/rng.py:18]: we need to do different things if its an enum compared to if its a list
```py
            if isinstance(rarity_weights, Rarity):
                self.weights = rarity_weights.weights

            elif isinstance(rarity_weights, list):
                self.weights = rarity_weights
```

**Review by GDWR — APPROVED** [review]

**Comment by GDWR** [comment]
Some conflicts then it should be good

**Review by GDWR — APPROVED** [review]

**Pull request merged**